[Week5] 고양이 장난감가게 유효성 추가하기, 회원 관리 만들기

JIY0UNG opened the pull request:
안녕하세요!!
User 인스턴스 생성과 수정에 대한 테스트 코드 작성 후 이를 통과하도록 구현하는 것까지 해보았습니다.
리뷰 부탁드립니다!

JIY0UNG committed:
Gradle과 Lombok의 버전을 변경하라

jdk와 호환되지 않는 문제가 발생하여 Gradle 버전을 7.3으로, Lombok 버전을 1.18.24로 변경했습니다.

JIY0UNG committed:
Windows, Intellij, Gradle, Java에 대한 gitignore을 추가하라

JIY0UNG committed:
User 인스턴스 생성에 대한 테스트 코드를 작성하라

JIY0UNG committed:
User 인스턴스 수정에 대한 테스트 코드를 작성하라

JIY0UNG committed:
User 인스턴스 생성, 수정에 대한 테스트 코드를 통과하도록 구현하라

ssisksl77 commented on app/src/test/java/com/codesoom/assignment/domain/UserTest.java:
```diff
@@ -0,0 +1,35 @@
+    }
+
+    @Test
+    void update() {
```
테스트에 대한 설명이 필요해보여요! 
한번 계층구조로 나눠서 테스트를 좀 더 설명해주시면 좋겠어요!

JIY0UNG replied:
모든 테스트를 통과하도록 구현한 과제를 기한 내에 제출하는 것이 이번 주의 최우선 목표라서,
추후에 리팩토링 하겠습니다!!

JIY0UNG committed:
POST /user 요청에 대한 컨트롤러 테스트 코드를 작성하라

JIY0UNG committed:
BadRequestException 클래스를 만들어라

유효하지 않은 속성으로 요청 시 BadRequestException을 발생시키기 위해 만들었습니다.

JIY0UNG committed:
ErrorAdvice에 BadRequest에 대한 handler메서드를 추가하라

한 개의 ErrorAdvice에서 모든 에러를 관리하기 위해 기존 클래스 명인 NotFoundErrorAdvice를 ErrorAdvice로 변경한 후 handleBadRequest 메서드를 추가했습니다.

JIY0UNG committed:
Gradle과 Lombok의 버전을 변경하라(다시 다운그레이드)

기존에 사용하던 jdk 18 버전과 호환되지 않아, Gradle과 Lombok의 버전을 업그레이드했었습니다. 그러나 이 때문에 Spring과 호환되지 않는 문제가 새로 발생하여, jdk 15 버전을 기준으로 모든 도구들이 서로 호환될 수 있도록 다시 버전 변경을 하였습니다.

JIY0UNG committed:
POST /user 요청에 대한 컨트롤러 테스트를 통과하도록 구현하라

JIY0UNG committed:
any() 메서드를 eq() 메서드로 변경하라

매개변수가 어떤 값이라도 상관없는 경우에는 any() 메서드를 쓰는 것이 맞지만, 특정한 값들을 가지는 객체가 매개변수로 주어져야 할 때는 eq() 메서드를 쓰는 것이 적절하기 때문에 변경하였습니다.

JIY0UNG committed:
UserNotFoundException 클래스를 만들고, ErrorAdvice에 추가하라

- 유효하지 않은 id로 요청 시 UserNotFoundException을 발생시키기 위해 예외 클래스를 만들었습니다.
- ErrorAdvice에서 해당 예외를 관리할 수 있도록 handleUserNotFound() 메서드를 추가하였습니다.

JIY0UNG committed:
PATCH /user 요청에 대한 컨트롤러 테스트 코드를 작성하라

JIY0UNG committed:
PATCH /user 요청에 대한 컨트롤러 테스트를 통과하도록 구현하라

JIY0UNG committed:
DELETE /user 요청에 대한 컨트롤러 테스트 코드를 작성하라

JIY0UNG committed:
JpaUserRepository에서 save()의 매개변수와 반환값을 User로 수정하라

save() 메서드가 매개변수로 Product를 받고 Product를 반환하고 있었다는 것을 발견하여 수정했습니다.

JIY0UNG committed:
DELETE /user 요청에 대한 컨트롤러 테스트를 통과하도록 구현하라

JIY0UNG committed:
User 생성에 대한 서비스 테스트 코드를 작성하라

JIY0UNG commented:
회원 생성, 수정, 삭제에 대한 컨트롤러 구현을 TDD로 해보았습니다! 리뷰 부탁드립니다😊

ssisksl77 commented on app/src/main/java/com/codesoom/assignment/dto/UserData.java:
```diff
@@ -0,0 +1,23 @@
+    @NotBlank
+    private String email;
+
+    @NotBlank
```
좋아요! 우리 한번 경험삼아 `@NotBlank` 와 동일하게 동작하는 커스텀 어노테이션을 만들어서 접목해보면 어떨까요?!

ssisksl77 commented on app/src/test/java/com/codesoom/assignment/controllers/UserControllerTest.java:
```diff
@@ -0,0 +1,183 @@
+
+        UserData userDataWithInvalidEmail = UserData.builder()
+                .name("김철수")
+                .email("")
+                .password("1111")
+                .build();
+
+        UserData userDataWithInvalidPassword = UserData.builder()
+                .name("김철수")
+                .email("kim@gmail.com")
+                .password("")
+                .build();
+
+        UserData newUserDataWithInvalidName = UserData.builder()
+                .name("")
+                .email("young@gmail.com")
+                .password("1234")
+                .build();
+
+        given(userService.create(any(UserData.class))).willReturn(user);
+        given(userService.create(eq(userDataWithInvalidName))).willThrow(new BadRequestException());
+        given(userService.create(eq(userDataWithInvalidEmail))).willThrow(new BadRequestException());
+        given(userService.create(eq(userDataWithInvalidPassword))).willThrow(new BadRequestException());
+
+        given(userService.update(eq(1L), any(UserData.class)))
+                .will(invocation -> {
+                    Long id = invocation.getArgument(0);
+                    UserData userData = invocation.getArgument(1);
+
+                    return User.builder()
+                            .id(1L)
+                            .name(userData.getName())
+                            .email(userData.getEmail())
+                            .password(userData.getPassword())
+                            .build();
+        });
+        given(userService.update(eq(1000L), any(UserData.class))).willThrow(new UserNotFoundException(1000L));
+        given(userService.update(eq(1L), eq(newUserDataWithInvalidName))).willThrow(new BadRequestException());
+
+        given(userService.delete(1000L)).willThrow(new UserNotFoundException(1000L));
```
목 설정은 각 테스트에 위치하는 것이 좋겠어요!

JIY0UNG replied:
네 수정하겠습니다!

JIY0UNG committed:
User 수정에 대한 서비스 테스트 코드를 작성하라

JIY0UNG committed:
User 삭제에 대한 서비스 테스트 코드를 작성하라

JIY0UNG committed:
DTO 클래스의 필드 중 id를 제거하라

DTO에는 id가 필요 없으므로 제거했습니다.

JIY0UNG committed:
DTO에 toString() 메서드를 추가하라

JIY0UNG committed:
ProductData의 imageUrl 필드에 대해 NotBlank 어노테이션을 붙여라

JIY0UNG committed:
ProductControllerTest 클래스의 create() 메서드를 제거하라

유효한 속성으로 create 하는 경우에 대한 테스트 메서드가 있으므로 제거하였습니다.

JIY0UNG committed:
UserServiceTest에서 불필요한 import문을 제거하라

Product 관련 import 문들은 필요 없으므로 제거하였습니다.

JIY0UNG committed:
ProductControllerTest에서 accept() 메서드를 제거하라

현재 사용 중인 Spring Boot 2.3.5 버전에서는 MediaType.APPLICATION_JSON_UTF8 이 사용되지 않기도 하고, contentType() 메서드에서 MediaType을 지정해주고 있으므로 필요 없다고 생각하여 제거했습니다.

JIY0UNG committed:
결과 테스트 시 어떤 값에 대한 테스트인지 명확히 알 수 있도록 변경하라

ProductControllerTest에서 기존에는 containsString으로 '결과값이 어떤 문자열을 가지고 있는지'를 테스트했었는데, 테스트하는 값을 좀 더 명확하게 표현하기 위해서 jsonPath("$.name").value()의 형식으로 변경하였습니다.

JIY0UNG committed:
HTTP 요청 시 보내는 데이터에 imageUrl을 추가하라

DTO에 imageUrl 필드도 있기 때문에 추가하였습니다.

JIY0UNG committed:
DTO의 toString() 메서드에 대한 테스트 코드를 작성하라

JIY0UNG committed:
DTO의 builder.toString() 에 대한 테스트 코드를 작성하라

JIY0UNG committed:
Domain의 기본 생성자에 대한 테스트 코드를 작성하라

JIY0UNG committed:
Domain의 builder.toString() 에 대한 테스트 코드를 작성하라

JIY0UNG commented:
오늘은 테스트 커버리지를 높이기 위한 작업들을 했습니다!

이제 이번 주 남은 기간 동안 아래 항목들을 할 예정입니다
- ErrorAdvice 커버리지 100% 달성시키기
- 계층 구조로 리팩토링
- Dozer를 사용하여 객체 매핑을 하도록 리팩토링

JIY0UNG committed:
ProductControllerTest feature에 imageUrl 속성을 추가하라

JIY0UNG committed:
DozerMapper 의존성을 추가하라

JIY0UNG committed:
App클래스의 불필요한 getGreeting() 메서드와 이에 대한 테스트 코드를 제거하라

JIY0UNG committed:
DozerMapper를 Bean으로 등록하라

JIY0UNG committed:
Dozer Mapper를 이용하여 DTO와 Domain을 매핑하라

매핑을 좀 더 간편하게 하기 위해 Dozer Mapper를 사용하였습니다.

JIY0UNG committed:
Product의 GET, POST 요청에 대한 컨트롤러 테스트를 계층형으로 리팩토링 하라

- GET: 유효한 id가 주어진 경우, 유효하지 않은 id가 주어진 경우에 대한 테스트 코드를 계층형으로 리팩토링 했습니다.
- POST: 유효한 속성을 가진 product가 주어진 경우에 대한 테스트 코드를 계층형으로 리팩토링 했습니다.

JIY0UNG commented:
오늘은 Dozer Mapper를 적용하고, 일부 컨트롤러 테스트 코드를 계층형으로 리팩토링 했습니다.
리뷰 부탁드립니다!

ssisksl77 commented on app/src/main/java/com/codesoom/assignment/App.java:
```diff
@@ -1,11 +1,19 @@
         SpringApplication.run(App.class, args);
     }
+
+    @Bean
+    public Mapper dozerMapper() {
+        return DozerBeanMapperBuilder.buildDefault();
+    }
```
빈에 등록하는 것은 아주 좋은 시도입니다!

조금 아쉬운 점은 빈을 등록하는 코드가 App 이라는 코드에 있다는 것입니다! 

분리를 해보면 좋을 것 같아요!

ssisksl77 commented:
dozer mapper가 아주 편하지요! 

Mapper 를 사용하면서 느낀 장,단점을 기술해주세요!

대부분의 기술은 도입하게 되면 장점과 단점(존재한다면!)을 공유하는 것이 일반적이에요! 지금부터 연습을 해보면 좋겠어요!

ssisksl77 commented:
커스텀 어노테이션을 만들어서 접목해보는 것이 실무에 많은 도움이 될거에요! 한번 기존 유효성검증 어노테이션을 커스텀 어노테이션으로 대체해보세요!

JIY0UNG committed:
ProductControllerTest를 계층형으로 리팩토링 하라

POST, PATCH, DELETE에 대한 컨트롤러 테스트 코드를 모두 계층형으로 리팩토링 했습니다.

JIY0UNG committed:
ProductData를 json으로 변환하는 코드를 별도의 메서드로 분리하라

ProductData를 json으로 변환하는 코드가 여러 테스트 코드에서 반복적으로 사용되므로 toJson() 메서드로 분리하였습니다.

JIY0UNG committed:
ErrorAdvice에서 handleBadRequest() 메서드를 제거하라

BadRequestException은 컨트롤러에서 직접 던져주는 예외가 아니므로 제거했습니다.

JIY0UNG committed:
App에 대한 테스트 코드를 작성하라

JIY0UNG committed:
App에 대한 테스트 코드를 통과하도록 getGreeting() 메서드를 추가하라

JIY0UNG committed:
UserControllerTest를 계층형으로 리팩토링 하라

ssisksl77 commented on app/src/test/java/com/codesoom/assignment/controllers/ProductControllerTest.java:
```diff
@@ -210,74 +184,202 @@ void test() throws Exception {
+        class Context_if_valid_id_given {
+            @BeforeEach
+            void setUp() {
+                given(productService.updateProduct(eq(ID), any(ProductData.class)))
+                        .will(invocation -> {
+                            Long id = invocation.getArgument(0);
+                            ProductData productData = invocation.getArgument(1);
+                            return Product.builder()
+                                    .id(id)
+                                    .name(productData.getName())
+                                    .maker(productData.getMaker())
+                                    .price(productData.getPrice())
+                                    .imageUrl(productData.getImageUrl())
+                                    .build();
+                        });
```
우리는 Context에 한가지 더 추가할 필요가 있습니다.

바로 given에 해당하는 코드들 또한 가정에 들어가기 때문이죠!

JIY0UNG committed:
회원 관련 요청 URL을 users로 변경하라

JIY0UNG committed:
회원 수정용 DTO를 통해 데이터 수정을 하도록 변경하라

기존에는 UserData로 회원 생성과 수정을 함께 했었는데, email 필드에 NotBlank 어노테이션이 붙어있어서 API 테스트 중 PATCH 부분이 통과되지 않았습니다.
따라서 email 필드에 NotBlank 어노테이션을 붙이지 않은 회원 수정용 DTO(UserUpdateData)를 별도로 만들어서 회원 생성용 DTO(UserCreateData)와 구분하였습니다.

JIY0UNG committed:
컨트롤러에 CrossOrigin 어노테이션을 추가하라

프론트와 백엔드가 서로 요청을 주고받을 수 있도록 CrossOrigin 설정을 해주었습니다.

JIY0UNG committed:
ProductData imageUrl 필드의 NotBlank 어노테이션을 제거하라

E2E 테스트 중 imageUrl을 넣지 않았을 경우에도 장난감 등록이 되어야 하는 테스트가 있는데, 이를 통과하기 위해서 NotBlank 어노테이션을 제거하였습니다.
